### ISSUE TITLE
NPE in ClassMirror.instanceMembers unless every immediate subclass is marked for reflection or superclassQuantifyCapability is specified

### ISSUE THREAD

**Issue by lassedamgaard** [dart-lang/reflectable#51]
This one had me scratching my head for a while.

Given the capabilities `instanceInvokeCapability, metadataCapability, nameCapability` and the class C extending class B that extends class A where class C is annotated with my capability spec I get an NPE in javascript when I do: 
```
final members = myReflectionCapability.reflect(instanceOfClassC).type.instanceMembers;
```

The error occurs in the `_variableMirror ` getter here because `_variableMirrorIndex` is null.
```
abstract class ImplicitAccessorMirrorImpl extends _DataCaching
    implements MethodMirror {
  final ReflectableImpl _reflector;
  final int _variableMirrorIndex;
  final int _reflectedTypeIndex;
  final int _dynamicReflectedTypeIndex;

  /// Index of this [ImplicitAccessorMirrorImpl] in `_data.memberMirrors`.
  final int _selfIndex;

  VariableMirrorImpl get _variableMirror =>
      _data.memberMirrors[_variableMirrorIndex];

  ImplicitAccessorMirrorImpl(
      this._reflector,
      this._variableMirrorIndex,
      this._reflectedTypeIndex,
      this._dynamicReflectedTypeIndex,
      this._selfIndex);

  int get kind => constants.kindFromEncoding(_variableMirror._descriptor);
```

This is the stacktrace from Chrome
```
2015-12-04 14:01:17.320 [FINER]  [HistoryManager] Reflecting on ServiceList
js_primitives.dart:31 Invalid argument (index): null

STACKTRACE:
Invalid argument (index): null
    at dart.wrapException (http://localhost:2001/main.dart.js:3414:17)
    at dart.ioore (http://localhost:2001/main.dart.js:3362:15)
    at ImplicitGetterMirrorImpl.dart.ImplicitGetterMirrorImpl.get$simpleName (http://localhost:2001/main.dart.js:49701:20)
    at NonGenericClassMirrorImpl.dart.ClassMirrorBase.get$instanceMembers (http://localhost:2001/main.dart.js:49446:51)
    at _ComponentReflector.dart._ComponentReflector.getStatefulProperties$1 (http://localhost:2001/main.dart.js:70686:101)
    at HistoryManager.dart.HistoryManager._handleComponentAttach$1 (http://localhost:2001/main.dart.js:70483:47)
    at HistoryManager__addEventListeners_closure.dart.HistoryManager__addEventListeners_closure.call$1 (http://localhost:2001/main.dart.js:70535:55)
    at ScopeStreamSubscription.dart.ScopeStreamSubscription._core_internal$_onData$1 (http://localhost:2001/main.dart.js:13733:44)
    at ScopeStream.dart.ScopeStream._fire$1 (http://localhost:2001/main.dart.js:13659:28)
    at dart._Streams.static._Streams_emit (http://localhost:2001/main.dart.js:13527:24)
```

If I add `superclassQuantifyCapability` it works but this seems like a bug because I actually only want to reflect on C. 

The error also does not occur or cause any warning in the VM. Also it doesn't occur if I leave out class A, meaning that it's just C extending B or if I add my reflection annotation to B.

**Comment by eernstg** [comment]
Currently debugging treatment of NoSuchMethodError, I suspect that this may be related to an issue that I discovered while doing that. Will dive into this when that other CL is settled.

**Comment by eernstg** [comment]
The CL https://codereview.chromium.org/1509673003/ contains a new test that confirms that commit 5a835be813daeac500ec3f7ccc8fef6e20c58e30 (the most recent now) already addresses this bug.

Thanks for the input!

**Comment by lassedamgaard** [comment]
I can confirm that the issue is fixed at 5a835be813daeac500ec3f7ccc8fef6e20c58e30. Thanks for the swift response!
